storage/blockdevice/COMPONENT_QSPIF/source/QSPIFBlockDevice.cpp

```diff
@@ -554,7 +554,7 @@ int QSPIFBlockDevice::get_erase_value() const
 /********************************/
 static PinName *generate_initialized_active_qspif_csel_arr()
 {
-    PinName *init_arr = new PinName[QSPIF_MAX_ACTIVE_FLASH_DEVICES];
+    static PinName init_arr[QSPIF_MAX_ACTIVE_FLASH_DEVICES];
```

Comment:
Is there a reason we can't use a `std::array` instead of a c-style array? Maybe we don't need value-semantics here as this is `static` anyway?

Reply:
Could you explain mean by value-semantics here?

As for `std::array`, I guess the benefits are like iterators and bound checks, etc. But the bound check can potentially throw `std::out_of_range` which we can't catch due to `-fno-exceptions`.

Reply:
By value-semantics I mean that the array itself will be treated as a value rather than decaying to a pointer to its first element. So, when passing it as an argument to a function we can pass the array by `const &` or by value instead of using pointer semantics. We don't have to pass a "size" argument with it. We don't have to keep track of the size at all. The size is actually part of the type. So, you could write a function that only takes an array of size N, which would fail to compile if we passed it an incorrectly sized array. It's generally good style to default to `std::array` if we have a fixed size array, rather than using the more error prone "C-style" array for those reasons. 

This was a question rather than a request for a change, anyway. As I mentioned, because this is static, maybe we don't need to copy or pass it as a parameter anywhere. It's good to at least consider using the more modern c++ constructs in as many cases as possible (they're generally stronger typed and safer/easier to use than the C equivalents), and I wasn't sure if we'd considered it.

Reply:
Totally agree. The limitation of the current implementation is, this function returns a pointer, which gets set as the value of `QSPIFBlockDevice::_active_qspif_flash_csel_arr` (a static pointer) so it gets back to C-style code again. But we're free to rework the internals of a class as long as it doesn't break any public API. I'd prefer making `QSPIFBlockDevice::_active_qspif_flash_csel_arr` a static array directly, so we won't lose size info. Will make the change.

Reply:
This would involve a bit of refactoring. I'll let other reviewers say if we want to do it in this PR.

Reply:
How much memory is this? I'd prefer to use the heap so this memory can be used for other purposes if needed. Static hogs the memory here, whether it's in use or not.

This is a static function, so we can't just free the memory in the QSPIFBlockDevice destructor, but it'd be good to understand how much memory is being used here (at least on one or two typical devices) and use that to help justify making this change. Please update the commit message "QSPIFBlockDevice: Use static memory for CSEL array" with this information.